### ISSUE TITLE
[inductor] Reshape across a sub-stick dim ([T,K]→[N,1]) aborts the layout pass

### ISSUE THREAD

**Issue by ani300** [torch-spyre/torch-spyre#3634]
# [inductor] Reshape across a sub-stick dim ([T,K]→[N,1]) aborts the layout pass

## Summary

Reshaping a `[T, K]` fp16 tensor with a partial stick dim (`K = 4 < 64`) to a flat
`[N=T*K, 1]` and using it as a broadcast operand in a pointwise multiply aborts the
torch-spyre layout pass with *no supported output layout found*.

This is the MoE router-weight scaling step: each routed row must be scaled by its
router weight `w`, which is naturally `[T, K]` (T tokens × K experts) and gets
flattened to `[N, 1]` to line up with the `[N, H]` routed-row buffer. That flatten
is what blocks.

## Repro

`repros/gemma4_moe/issues/p1_reshape_substick_abort.py` (no model load, no real weights):

```python
import torch, torch_spyre
T, K, H = 64, 4, 2816
N = T * K  # 256

def bug(w_tk, x_nh):
    return x_nh * w_tk.reshape(N, 1)   # [T,K] -> [N,1] across the K=4 partial stick

def control(w_n1, x_nh):
    return x_nh * w_n1                 # device-native [N,1], never [T,K]-shaped

w_tk = torch.randn(T, K, dtype=torch.float16).to("spyre")
x_nh = torch.randn(N, H, dtype=torch.float16).to("spyre")
w_n1 = torch.randn(N, 1, dtype=torch.float16).to("spyre")

torch.compile(bug,     dynamic=False)(w_tk, x_nh).cpu()   # ABORTS
torch.compile(control, dynamic=False)(w_n1, x_nh).cpu()   # compiles fine
```

## Observed (on-device, 2026-08-10)

```
bug     : ABORT  Unsupported: Spyre backend does not support:
          Multi-arg pointwise (buf0): no supported output layout found
          with size=[256, 2816] and coordinates=[d0, d1]
control : COMPILED OK
```

The **control** — a device-native `[N,1]` operand that is never `[T,K]`-shaped —
compiles, so the trigger is specifically the `[T,K]→[N]` reshape across the partial
stick, not the `[N,H] * [N,1]` broadcast itself.

## Root cause (analysis)

`K = 4` occupies a stick padded to 64. Reading the `[T, K]` buffer with a flat `[N]`
index produces the cross-stick access expression `Mod(d0, 4) + floor(d0/4)` (mod **4**,
not mod 64). `is_stick_expr_offset_free` rejects it, so `_multi_arg_pointwise_layouts`
finds no supported output layout and raises at `propagate_layouts.py:1099` (gate at
`:985`). The comment at `propagate_layouts.py:987-989` anticipates inserting a
restickify for exactly this case, but none is emitted, and `optimize_restickify.py:183`
(`AllSameNode`) asserts `out_layouts` is non-empty.

## Impact

Blocks the MoE row-weight scaling regardless of how routing is structured (per-row
loop or per-expert grouping) — any path that flattens the `[T,K]` router weights to
`[N,·]` hits this.

## Suggested fix

Emit a `spyre.restickify` (HBM round-trip) for a reshape that crosses a sub-stick
dimension, as the `:987-989` comment already anticipates, so the flattened operand
lands in a stick-offset-free layout before the pointwise consumer.

## Labels

bug, inductor, tensor layout

**Comment by mudhakar** [comment]
@ani300 is this a possible workaround:

gather the routed rows with the 2-D index so they are born [T,K,H], scale in 3-D, and flatten after the multiply.

### WORKS — compiles, cos = 0.999994 vs CPU, on-device topk router included
w, sel = torch.topk(logits, K, dim=-1)             # [T,K]
rows   = table[sel]                                # [T,K,H]  <-- 2-D index
out    = (rows * w.unsqueeze(-1)).reshape(N, H)    # flatten the RESULT

### ABORTS — this issue
rows = table[sel.reshape(N)]                       # [N,H]
out  = rows * w.reshape(N, 1)

The [T,K,H] has to come from the gather, not from a later view — table[sel.reshape(N)].reshape(T,K,H) * w.unsqueeze(-1) aborts the same way (the reshapes fuse back to the flat form).

**Comment by ani300** [comment]
No longer needed for MoE enablement
